### PR TITLE
[BACKLOG-12457] Added the notion of view state to pentaho.visual.base…

### PR DESCRIPTION
**Pull request by carlosrusso**
….View

@pentaho/millenniumfalcon Please review

### CODE REVIEW

File: test-js/unit/pentaho/visual/base/View.Spec.js [hunk 1]
```diff
@@ -1164,6 +1171,53 @@ define([
...
+
+      it("should return the state of the view", function() {
+
+        var state = {};
+        var view = new DerivedView(document.createElement("div"), model, state);
+
+        expect(view.getState()).toBe(state);
+      });
+
+    }); // #toJSON
```
Comment (davidmsantos90) [inline test-js/unit/pentaho/visual/base/View.Spec.js]: this should be // #getState

File: package-res/resources/web/pentaho/visual/base/View.js [hunk 2]
```diff
@@ -80,8 +80,9 @@ define([
      * @description Initializes a `View` instance.
      * @param {!DOMElement} domContainer - The container element.
      * @param {!pentaho.visual.base.Model} model - The visualization model.
+     * @param {state} model - The initial state of the view.
```
Comment (davidmsantos90) [inline package-res/resources/web/pentaho/visual/base/View.js]: you have the name of the parameter inside the type slot.

@param {type} state - ...

File: package-res/resources/web/pentaho/visual/base/View.js [hunk 3]
```diff
@@ -104,6 +105,15 @@ define([
       this.__model = model;
 
       /**
+       * The state of the view.
+       *
+       * @type {object}
+       * @readOnly
+       * @private
+       */
+      this.__state = state;
```
Comment (dcleao) [inline package-res/resources/web/pentaho/visual/base/View.js]: I don't think state should be stored by the base class — it is "abstract" in this respect.
It is up to specific view sub-classes to "parse" it and store it somehow.

File: package-res/resources/web/pentaho/visual/base/View.js [hunk 4]
```diff
@@ -80,8 +80,9 @@ define([
      * @description Initializes a `View` instance.
      * @param {!DOMElement} domContainer - The container element.
      * @param {!pentaho.visual.base.Model} model - The visualization model.
+     * @param {object} state - The initial state of the view.
```
Comment (dcleao) [inline package-res/resources/web/pentaho/visual/base/View.js]: `@param {Object} [state] - The initial state.`

File: package-res/resources/web/pentaho/visual/base/View.js [hunk 5]
```diff
@@ -972,6 +982,26 @@ define([
      * @see pentaho.visual.base.View#_updateAll
      */
     // endregion
+  })
+  .implement(/** @lends pentaho.visual.base.View# */{
+
+    /**
```
Comment (dcleao) [inline package-res/resources/web/pentaho/visual/base/View.js]:
```js
/**
 * Gets the state of the view as a plain object, if any.
 *
 * This method simply calls the [getState]{@link pentaho.visual.base.View#getState} method
 * and provides support for `JSON.stringify`.
 *
 * @return {Object} The state of the view, or `null`, if none.
 *
 * @see pentaho.visual.base.View#getState
 */
```

Also, please, call `getState` from this method, and not the other way round.

File: package-res/resources/web/pentaho/visual/base/View.js [hunk 6]
```diff
@@ -972,6 +982,26 @@ define([
...
+    toJSON: function() {
+      return this.__state || null;
+    },
+
+    /**
+     * Gets the state of the view.
+     *
+     * @return {object}
+     */
+    getState: function() {
```
Comment (dcleao) [inline package-res/resources/web/pentaho/visual/base/View.js]:
```js
/**
 * Gets the state of the view as a plain object, if any.
 *
 * The object returned by this method can later be given to the view's constructor, in argument `state`.
 *
 * The returned object must be JSON serializable (must have no cycles, and can only contain numbers, booleans, strings, plain objects and arrays).
 *
 * This implementation simply returns `null`. Override this method to expose a view's specific state. When overriding, take care to respect a state object that may be returned by the base class implementation.
 *
 * @return {Object} The state of the view, or `null`, if none. 
 * @see pentaho.visual.base.View#toJSON
 */
```

Please, just return `null`.

**Pull request merged**